Mathlib/NumberTheory/ModularForms/EisensteinSeries/UniformConvergence.lean

```diff
@@ -0,0 +1,364 @@
+/-
+Copyright (c) 2024 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import Mathlib.Analysis.Complex.UpperHalfPlane.Basic
+import Mathlib.Data.Int.Interval
+import Mathlib.Analysis.SpecialFunctions.Pow.Complex
+import Mathlib.Analysis.SpecialFunctions.Pow.Real
+import Mathlib.NumberTheory.ModularForms.EisensteinSeries.Basic
+import Mathlib.Analysis.Complex.UpperHalfPlane.Metric
+import Mathlib.Analysis.NormedSpace.FunctionSeries
+import Mathlib.Analysis.PSeries
+import Mathlib.NumberTheory.ModularForms.EisensteinSeries.Finset_Decomposition
+
+/-!
+# Uniform convergence of Eisenstein series
+
+We show that `eis` converges locally uniformly on `ℍ` to the Eisenstein series `E` of weight `k`
+and level `Γ(N)` with congruence condition `a : Fin 2 → ZMod N`.
+-/
+
+noncomputable section
+
+open Complex Filter UpperHalfPlane Set
+
+open scoped BigOperators NNReal Classical Filter Matrix UpperHalfPlane Complex
+
+namespace EisensteinSeries
+
+section bounding_functions
+
+/-- Auxilary function used for bounding Eisentein series-/
+def r1 (z : ℍ) : ℝ := ((z.im ^ (2 : ℕ)) / (z.re ^ (2 : ℕ) + z.im ^ (2 : ℕ)))
+
+lemma r1' (z : ℍ) : r1 z = 1/((z.re / z.im) ^ (2 : ℕ) + 1) := by
+  field_simp [r1, im_pos z]
+
+theorem r1_pos (z : ℍ) : 0 < r1 z := by
+  have H2 : 0 < (z.re ^ (2 : ℕ) + z.im ^ (2 : ℕ)) := by
+    apply_rules [pow_pos, add_pos_of_nonneg_of_pos, pow_two_nonneg, z.2]
+  exact div_pos (pow_pos z.im_pos 2) H2
+
+/-- This function is used to give an upper bound on Eisenstein series-/
+def r (z : ℍ) : ℝ := min (z.im) (Real.sqrt (r1 z))
+
+theorem r_pos (z : ℍ) : 0 < r z := by
+  simp only [r, lt_min_iff, im_pos, Real.sqrt_pos, r1_pos, and_self]
+
+lemma r_mul_pos_pos (k : ℕ) (z : ℍ) (n : ℝ) (hn : 0 < n) :
```

Comment:
```suggestion
lemma r_mul_pos_pos (k : ℕ) (z : ℍ) {n : ℝ} (hn : 0 < n) :
```
This seems like a strange lemma, are you just saying that the product is different from zero?

Reply:
So I do actually need that it is positive. But thats a good point, at one point I needed this several times which is why I added it. But I think I've now managed to remove most of the need for it, so I'll remove the lemma.